File: lazypx4/mavlink/calibration.py
"""Sensor calibration (MAV_CMD_PREFLIGHT_CALIBRATION).

PX4 runs gyro / accel / level / magnetometer / baro calibration entirely
on-board: we send one COMMAND_LONG and then follow the "[cal] ..." STATUSTEXT
messages it streams back for progress and for the "place the vehicle
nose-down" style prompts. Accel and mag need the operator to physically move
the airframe; gyro / level / baro just need it to sit still. Must be done
DISARMED.
"""

from __future__ import annotations

import re
import time

from ..config import (
    CALIBRATION_LABELS,
    CALIBRATION_PARAMS,
    MAV_CMD_PREFLIGHT_CALIBRATION,
)
from ..eventlog import log_command, log_error, log_warn
from ..state import state
from ..util import clamp, safe_int
from .connection import vehicle_ready


def handle_calibration_text(text, now):
    """Fold a "[cal]" STATUSTEXT line into the calibration state."""
    lower = text.lower()

    with state.lock:
        state.cal_last_message = text
        state.cal_last_message_at = now
        state.cal_log.append(text)

        match = re.search(r"progress\s*<?\s*(\d+)", lower)
        if match:
            state.cal_progress = clamp(safe_int(match.group(1)), 0, 100)

        if "calibration started" in lower:
            state.cal_active = True
            state.cal_progress = 0
            state.cal_result = ""
            state.cal_started_at = now
        elif "calibration done" in lower:
            state.cal_active = False
            state.cal_progress = 100
            state.cal_result = "DONE"
        elif "calibration failed" in lower:
            state.cal_active = False
            state.cal_result = "FAILED"
        elif "calibration cancel" in lower:
            state.cal_active = False
            state.cal_result = "CANCELLED"


def send_calibration(master, cal_type):
    if not vehicle_ready(master):
        return False

    params = CALIBRATION_PARAMS.get(cal_type)
    if params is None:
        log_error(f"Unknown calibration type: {cal_type}")
        return False

    with state.lock:
        if state.armed:
            log_error("Calibration refused: vehicle is ARMED")
            return False
        if state.cal_active:
            log_warn("A calibration is already running")
            return False

    try:
        master.mav.command_long_send(
            master.target_system,
            master.target_component,
            MAV_CMD_PREFLIGHT_CALIBRATION,
            0,
            *[float(p) for p in params],
        )
    except Exception as exc:
        log_error(f"Failed to send calibration command: {exc}")
        return False

    with state.lock:
        now = time.monotonic()
        state.cal_active = True
        state.cal_type = cal_type
        state.cal_progress = 0
        state.cal_result = ""
        state.cal_started_at = now
        state.cal_sent_at = now
        state.cal_ack = ""
        state.cal_ack_at = 0.0
        state.cal_statustext_at_start = state.statustext_count
        state.cal_event_at_start = state.event_count
        state.cal_last_message = ""
        state.cal_log.clear()

    log_command(
        f"PREFLIGHT_CALIBRATION SENT: {CALIBRATION_LABELS.get(cal_type, cal_type)} "
        f"(sys {master.target_system} comp {master.target_component})"
    )
    return True


def cancel_calibration(master):
    if master is None:
        return False

    try:
        master.mav.command_long_send(
            master.target_system,
            master.target_component,
            MAV_CMD_PREFLIGHT_CALIBRATION,
            0,
            0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        )
    except Exception as exc:
        log_error(f"Failed to cancel calibration: {exc}")
        return False

    with state.lock:
        state.cal_active = False
        state.cal_result = "CANCELLED"

    log_command("Calibration CANCEL sent")
    return True


def check_calibration():
    """Give up on a calibration that PX4 never acknowledged or started.

    A real calibration always produces either a COMMAND_ACK or a stream of
    "[cal]" STATUSTEXT messages within a few seconds. If neither happens the
    command did not land - clear the "RUNNING" state so the screen is usable
    again instead of hanging forever.
    """
    now = time.monotonic()

    with state.lock:
        if not state.cal_active:
            return

        sent_at = state.cal_sent_at
        have_ack = bool(state.cal_ack)
        have_log = bool(state.cal_log)

        if not sent_at:
            return
        if have_ack or have_log:
            return
        if now - sent_at < 25.0:
            return

        state.cal_active = False
        state.cal_result = "NO RESPONSE"

    log_error(
        "Calibration: PX4 sent no ACK and no [cal] messages in 25 s - "
        "command was not accepted. Check [g] event log and PX4 firmware."
    )

File: src/mav/calibration.rs
//! Sensor calibration (MAV_CMD_PREFLIGHT_CALIBRATION).
//!
//! PX4 runs gyro / accel / level / mag / baro calibration on-board: we send
//! one command and follow the "[cal] ..." STATUSTEXT stream for progress and
//! the "place the vehicle nose-down" style prompts. Must be done DISARMED.

use mavlink::dialects::development::MavCmd;

use crate::config::{calibration_label, calibration_params};
use crate::link::Link;
use crate::mav::commands::vehicle_ready;
use crate::mav::handlers::push_bounded;
use crate::state::{self, Shared, State, now};

/// Fold a "[cal]" STATUSTEXT line into the calibration state.
pub fn handle_calibration_text(st: &mut State, text: &str) {
    let lower = text.to_lowercase();
    st.cal_last_message = text.to_string();
    push_bounded(&mut st.cal_log, text.to_string(), 14);

    // "progress <NN>"
    if let Some(pos) = lower.find("progress") {
        let digits: String = lower[pos + 8..]
            .trim_start_matches(|c: char| c.is_whitespace() || c == '<')
            .chars()
            .take_while(char::is_ascii_digit)
            .collect();
        if let Ok(p) = digits.parse::<i32>() {
            st.cal_progress = p.clamp(0, 100);
        }
    }

    if lower.contains("calibration started") {
        st.cal_active = true;
        st.cal_progress = 0;
        st.cal_result.clear();
        st.cal_started_at = now();
    } else if lower.contains("calibration done") {
        st.cal_active = false;
        st.cal_progress = 100;
        st.cal_result = "DONE".into();
    } else if lower.contains("calibration failed") {
        st.cal_active = false;
        st.cal_result = "FAILED".into();
    } else if lower.contains("calibration cancel") {
        st.cal_active = false;
        st.cal_result = "CANCELLED".into();
    }
}

pub fn send_calibration(link: &Link, shared: &Shared, kind: &str) -> bool {
    let mut st = state::lock(shared);
    if !vehicle_ready(&mut st, link) {
        return false;
    }
    let Some(params) = calibration_params(kind) else {
        st.error(format!("Unknown calibration type: {kind}"));
        return false;
    };
    if st.armed {
        st.error("Calibration refused: vehicle is ARMED");
        return false;
    }
    if st.cal_active {
        st.warn("A calibration is already running");
        return false;
    }
    if !link.command_long(MavCmd::MAV_CMD_PREFLIGHT_CALIBRATION, params) {
        st.error("Failed to send calibration command");
        return false;
    }

    let t = now();
    st.cal_active = true;
    st.cal_type = kind.to_string();
    st.cal_progress = 0;
    st.cal_result.clear();
    st.cal_started_at = t;
    st.cal_sent_at = t;
    st.cal_ack.clear();
    st.cal_last_message.clear();
    st.cal_log.clear();

    let (sys, comp) = link.target();
    st.command(format!(
        "PREFLIGHT_CALIBRATION SENT: {} (sys {sys} comp {comp})",
        calibration_label(kind)
    ));
    true
}

pub fn cancel_calibration(link: &Link, shared: &Shared) -> bool {
    let mut st = state::lock(shared);
    if !link.command_long(MavCmd::MAV_CMD_PREFLIGHT_CALIBRATION, [0.; 7]) {
        st.error("Failed to cancel calibration");
        return false;
    }
    st.cal_active = false;
    st.cal_result = "CANCELLED".into();
    st.command("Calibration CANCEL sent");
    true
}

/// A real calibration always produces a COMMAND_ACK or "[cal]" text within
/// seconds; if neither arrives the command did not land, so clear RUNNING.
pub fn check_calibration(st: &mut State) {
    if !st.cal_active || st.cal_sent_at == 0.0 {
        return;
    }
    if !st.cal_ack.is_empty() || !st.cal_log.is_empty() || now() - st.cal_sent_at < 25.0 {
        return;
    }
    st.cal_active = false;
    st.cal_result = "NO RESPONSE".into();
    st.error(
        "Calibration: PX4 sent no ACK and no [cal] messages in 25 s - \
         command was not accepted. Check [g] event log and PX4 firmware.",
    );
}
